File: 22-generate-parentheses/solution.ts
function generateParenthesis(n: number): string[] {
  function backtrack(
    currentPermutation: string,
    solution: string[],
    leftParentheses: number,
    rightParentheses: number,
    n: number
  ) {
    if (currentPermutation.length === n * 2) {
      solution.push(currentPermutation);
      return;
    }
    if (leftParentheses < n) {
      backtrack(
        currentPermutation + "(",
        solution,
        leftParentheses + 1,
        rightParentheses,
        n
      );
    }
    if (rightParentheses < leftParentheses) {
      backtrack(
        currentPermutation + ")",
        solution,
        leftParentheses,
        rightParentheses + 1,
        n
      );
    }
  }
  const solution = [];
  backtrack("", solution, 0, 0, n);
  return solution;
}

console.log(generateParenthesis(4));

File: 22-generate-parentheses/solution.rs
impl Solution {
  fn generate_parenthesis(n: i32) -> Vec<String> {
    let mut solution: Vec<String> = Vec::new();
    let permutation = String::new();
    backtrack(&permutation, &mut solution, 0, 0, n);
    solution
  }
  
  fn backtrack(permutation: &String, solution: &mut Vec<String>, left: i32, right: i32, n: i32) {
    if permutation.len() as i32 == n * 2 {
      solution.push(*permutation);
      return;
    }
    if left < n {
      backtrack(&(permutation.to_owned() + "("), solution, left + 1, right, n);
    }
    if right < left {
      backtrack(&(permutation.to_owned() + ")"), solution, left, right + 1, n);
    }
  }
}

fn main() {
  Solution::generate_parenthesis(3);
}